src/muz/base/rule_properties.cpp

```diff
@@ -256,18 +258,62 @@ bool rule_properties::check_accessor(app* n) {
         if (!use_list.contains(e))
             return false;
         for (expr* parent : use_list[e]) {
-            if (!parent)
-                return false; // top-level expressions are not guarded
-            if (is_recognizer(parent)) 
-                continue;
-            if (m.is_ite(parent) && to_app(parent)->get_arg(1) == e && is_recognizer(to_app(parent)->get_arg(0))) 
+            if (!parent) { // top-level expression
+                // check if n is an unguarded "else" branch
+                ptr_vector<func_decl> ctors;
+                for (auto *rc : recognizers)
+                    if (to_app(rc)->get_arg(0) == n->get_arg(0))
+                        ctors.push_back(m_dt.get_recognizer_constructor(to_app(rc)->get_decl()));
+                ptr_vector<func_decl> diff;
+                for (auto *dtc : *m_dt.get_datatype_constructors(s))
+                    if (std::find(ctors.begin(), ctors.end(), dtc) == ctors.end())
+                        diff.push_back(dtc);
+                // the only unguarded constructor for s is c:
+                // all the others are guarded and we are in an "else" branch so the accessor is safe
+                if (diff.size() == 1 && diff[0] == c)
+                    continue;
+                return false; // the accessor is not safe
+            }
+            if (is_recognizer_base(parent))
                 continue;
-            todo.push_back(parent);                
+            if (m.is_ite(parent)) {
+                if (to_app(parent)->get_arg(1) == e && is_recognizer(to_app(parent)->get_arg(0)))
+                    continue; // e is guarded
+                else
+                    if (m_dt.is_recognizer(to_app(parent)->get_arg(0)))
+                        recognizers.push_back(to_app(parent)->get_arg(0)); // e is in the else branch of the ite
+            }
+            if (m.is_or(parent)) {
+                for (expr *arg : *to_app(parent)) {
+                    // if all branches are guarded the remaining one might be safe
+                    if (m_dt.is_recognizer(arg))
+                        recognizers.push_back(arg);
+                    // if one branch is not(recognizer) then the accessor is safe
+                    if (m.is_not(arg)) {
+                        if (is_recognizer(to_app(arg)->get_arg(0)))
+                            goto _continue;
```

Comment:
I need to skip the `todo.push_back`, so just `continue` from the inner for is not enough.
I'm not really happy about using `goto`, to remove it I could:
1. use `std::any_of` and `continue`
   ```c++
   if (m.is_or(parent)) {
       for (expr* arg : *to_app(parent))
           add_recognizer(arg);
       if (std::any_of(to_app(parent)->begin(), to_app(parent)->end(),
                       [&](expr *arg) { return m.is_not(arg, arg) && is_recognizer(arg); }))
           continue;
   }
   ```
   But this would loop twice over args
2. use a boolean flag and then continue if the flag is true
3. include the "or not" case in `is_recognizer`

Probably (1) is the cleaner one, even if it requires an additional loop/lambda.
I noticed that `goto` is used more than `std::any_of` in the codebase, so let me know which one you prefer for consistency.